File: src/BattleSim/Accuracy.lua
-- Gen1 accuracy: the 1/256 miss, written down on purpose.
--
-- Accuracy is an 8-bit value in 1..255 and the roll is a byte in 0..255, so
-- there are 256 rolls for at most 255 hitting values and a "100%" move misses
-- once in 256 tries.  That is a bug in the original and it is reproduced here,
-- because a mediated battle that quietly fixed it would resolve differently
-- from the same battle fought offline -- and players notice the fix, not the
-- faithfulness.
--
--   effectiveAccuracy = clamp(floor(floor(acc * accMod/100) * evaMod/100), 1, 255)
--   hit               = roll < effectiveAccuracy
--
-- The two modifier steps truncate separately, matching the original's
-- successive 8-bit multiplies, and the clamp is what stops a stack of
-- accuracy drops from reaching an unmissable 0 or a boost from reaching a
-- never-missing 256.
--
-- `alwaysHits` bypasses the roll entirely and reports no effective accuracy at
-- all -- there is no number to report, and returning 255 instead would be a
-- lie a caller could act on (it would still miss one time in 256).
--
-- Nothing here raises; no love, no engine modules, no mod facade.

local need = ...            -- entry-chunk shape; this module resolves nothing

local M = {}

local floor, max, min = math.floor, math.max, math.min

M.MIN = 1
M.MAX = 255

local function int(value, fallback)
  local n = tonumber(value)
  if not n or n ~= n then return fallback end
  return floor(n)
end

-- accuracy: 0..255 (the move's own value)
-- accuracyMod / evasionMod: percents, 100 being unmodified
function M.effective(accuracy, accuracyMod, evasionMod)
  local acc = max(0, int(accuracy, M.MAX))
  local value = floor(acc * max(0, int(accuracyMod, 100)) / 100)
  value = floor(value * max(0, int(evasionMod, 100)) / 100)
  return min(M.MAX, max(M.MIN, value))
end

-- Returns hit (boolean) and the effective accuracy it was decided against,
-- which is nil for a move that cannot miss.  Two returns rather than a table
-- because the answer callers want is almost always just the boolean.
--
-- opts: { accuracyMod, evasionMod, alwaysHits }
function M.hit(moveAccuracy, roll, opts)
  opts = opts or {}
  if opts.alwaysHits then return true, nil end
  local effective = M.effective(moveAccuracy, opts.accuracyMod, opts.evasionMod)
  return int(roll, 0) < effective, effective
end

return M

File: server/lib/battle/Accuracy.js
'use strict';

/*
 * Gen1 accuracy: does the move connect.
 *
 * Node half of src/BattleSim's accuracy step; the parity contract is the same
 * one Damage.js opens with.
 *
 * Accuracy is a 0..255 byte, not a percent, and the roll it is compared
 * against has 256 values. Since the effective accuracy is clamped to 255, a
 * move players call "100%" still misses on roll 255 -- the famous 1-in-256.
 * That is reproduced deliberately: a hub that quietly fixed it would disagree
 * with every other Gen1 implementation the player has ever seen, and would
 * disagree with the Lua twin the moment the twin did not fix it too.
 *
 * The modifiers are percents applied one at a time and truncated in between,
 * so an accuracy boost followed by an evasion boost is not commutative with
 * itself at every input. Order here is accuracy first, then evasion.
 */

const MIN = 1;
const MAX = 255;

const idiv = (a, b) => Math.floor(a / b);

const clamp = (v, lo, hi) => (v < lo ? lo : v > hi ? hi : v);

// Truncating between the two modifiers is the point -- do not fold them into
// one multiply.
function effective(accuracy, accuracyMod, evasionMod) {
  const withAcc = idiv(accuracy * (accuracyMod === undefined ? 100 : accuracyMod), 100);
  const withEva = idiv(withAcc * (evasionMod === undefined ? 100 : evasionMod), 100);
  return clamp(withEva, MIN, MAX);
}

/*
 * hit({ accuracy, accuracyMod, evasionMod, alwaysHits, roll })
 *
 * `alwaysHits` skips the roll entirely and reports a null effective accuracy:
 * a move that cannot miss has no accuracy to report, and returning 255 would
 * invite a caller to roll against it and reintroduce the 1-in-256.
 */
function hit(input) {
  if (input.alwaysHits) {
    return { effectiveAccuracy: null, hit: true };
  }
  const effectiveAccuracy = effective(input.accuracy, input.accuracyMod, input.evasionMod);
  return { effectiveAccuracy, hit: input.roll < effectiveAccuracy };
}

module.exports = { hit, effective, MIN, MAX };
